### DISCUSSION TITLE
How to close a loop with a tilted camera?

### DISCUSSION THREAD

**Discussion by udaykiran2021** [stella-cv/stella_vslam#218] [Q&A] [answered]
Hello,

Thanks for your great effort.

I have a question.
Is there a way that we can give rgbd camera extrinsic parameters as input to openvslam? Ex: camera angle.
If yes, could you please specify which source file needs to be modified?

Problem faced:
When the camera is tilted downward with an angle, the localization result is not good.

**Comment by ymd-stella** [comment, marked as answer]
> rgbd camera

What camera are you using?

> When the camera is tilted downward with an angle, the localization result is not good.

If it doesn't work fine, the first thing to do is to investigate the cause. We should not complicate the problem by adding new factors.
Can you provide config file and video input/output of OpenVSLAM?

**Reply by udaykiran2021** [reply]
We are using realsense d455.

As for config file: need to get permission to share.

**Reply by udaykiran2021** [reply]
I have extracted the intrinsic parameter values from d455 rosbag file and modified the example rgbd config.yaml. Please find the details below:

Camera:
  name: "D455"
  setup: "RGBD"
  model: "perspective"

  fx: 641.3826293945312
  fy: 641.3826293945312
  cx: 648.348876953125
  cy: 362.63299560546875

  k1: 0.0
  k2: 0.0
  p1: 0.0
  p2: 0.0
  k3: 0.0

  fps: 30.0
  cols: 1280
  rows: 720
  focal_x_baseline: 60.931349792

  color_order: "RGB"

**Reply by udaykiran2021** [reply]
Is there any issue in above config file?

**Reply by ymd-stella** [reply]
There seems to be no problem with the range of configs you are showing. (Except for the incorrect indentation)
What are the values of the following parameters?

- max_num_keypoints
- ini_max_num_keypoints
- depthmap_factor
- depth_threshold

Are you using openvslam_ros?
Is it difficult to provide video?

**Reply by udaykiran2021** [reply]
There is no change from the example config file except depthmap_factor.
```
Tracking:
  max_num_keypoints: 1000
  ini_max_num_keypoints: 2000
  depth_threshold: 40.0
  depthmap_factor: 1000.0
```

**Additional information:**
Case 1:
When camera mounted with some angle facing downwards:
During Map Creation (**Primary issue or concern**):
When the video is loop(start and end point is same), map created is distorted. Meaning, there is huge gap between start and end point in z axis when observed in socket viewer.
Tracking lost did not occur.

During Localization (Secondary issue, but can be accepted):
The tracking is lost during localization at very few points.

Case 2:
When camera mounted horizontal (parallel to floor):
During Map Creation:
No distortion while creating map. Both start and end point matched in all 3 axes.
Tracking lost did not occur.

During Localization:
Tracking lost at some point (but can be accepted).

Issue observed in non ros solution, but soon we will use openvslam-ros.

I am afraid, video can not be shared.

**Reply by ymd-stella** [reply]
I don't know how many degrees you are tilting the camera, but are you tilting the camera so far that no horizontal landmarks are visible?
In such a situation, matching between keyframes over long distances will not work, so it is not surprising that performance degrades.

Why do you want to tilt the camera?

**Reply by udaykiran2021** [reply]
RealSense is used for OpenVSLAM and Freespace detection, and Freespace should face downward as much as possible (about 1m away from feet)

Tilt angle = around 15deg.
I understand that fewer landmarks will result in less accuracy with tilted camera, but my problem is that I want to eliminate the height gap in z-axis when the camera is tilted.
I would like to reduce the z axis gap between start and end point which is same in my case.

**Reply by ymd-stella** [reply]
First, I recommend that you prepare one camera for localization and one for free space detection. In general, the camera for free space detection needs to be placed so that nearby objects are visible, while the camera for localization needs to be placed so that nearby objects are not visible. This means that the camera for localization should be placed higher.

Next, in a forced and practical way, place a landmark on the ground at the starting point. If OpenVSLAM can detect the loop, the loop closure can close the loop.

Or, although it may be difficult. Think about a way to increase the field of view. The field of view of the D455 is a little too narrow to be used for Visual SLAM. You might be able to stitch together images from multiple cameras and input them as an equirectangular format image.

**Reply by udaykiran2021** [reply]
Thank you so much for your suggestion. Let me discuss this and will get back.

**Reply by udaykiran2021** [reply]
> Next, in a forced and practical way, place a landmark on the ground at the starting point. If OpenVSLAM can detect the loop, the loop closure can close the loop.

We decided to go ahead with this suggestion and check if there is any improvement.
